// @agentsid/sdk/dist/errors.js
/**
 * AgentsID error classes
 */
export class AgentsIDError extends Error {
    code;
    statusCode;
    constructor(message, code, statusCode) {
        super(message);
        this.code = code;
        this.statusCode = statusCode;
        this.name = "AgentsIDError";
    }
}
export class AuthenticationError extends AgentsIDError {
    constructor(message = "Invalid or missing API key") {
        super(message, "AUTH_ERROR", 401);
        this.name = "AuthenticationError";
    }
}
export class PermissionDeniedError extends AgentsIDError {
    tool;
    reason;
    constructor(tool, reason) {
        super(`Permission denied for tool "${tool}": ${reason}`, "PERMISSION_DENIED", 403);
        this.tool = tool;
        this.reason = reason;
        this.name = "PermissionDeniedError";
    }
}
export class TokenExpiredError extends AgentsIDError {
    constructor() {
        super("Agent token has expired", "TOKEN_EXPIRED", 401);
        this.name = "TokenExpiredError";
    }
}
export class TokenRevokedError extends AgentsIDError {
    constructor() {
        super("Agent token has been revoked", "TOKEN_REVOKED", 401);
        this.name = "TokenRevokedError";
    }
}
export class AgentNotFoundError extends AgentsIDError {
    constructor(agentId) {
        super(`Agent not found: ${agentId}`, "AGENT_NOT_FOUND", 404);
        this.name = "AgentNotFoundError";
    }
}
